// mamba-ui/lib/mam-table/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var React = _interopRequireWildcard(require("react"));

var _table = _interopRequireDefault(require("antd/es/table"));

var _divider = _interopRequireDefault(require("antd/es/divider"));

var _popconfirm = _interopRequireDefault(require("antd/es/popconfirm"));

var _isBoolean = _interopRequireDefault(require("lodash/isBoolean"));

var _getComponentProps3 = _interopRequireDefault(require("../_utils/getComponentProps"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var MamTable =
/*#__PURE__*/
function (_React$Component) {
  _inherits(MamTable, _React$Component);

  function MamTable(props) {
    var _this;

    _classCallCheck(this, MamTable);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MamTable).call(this, props));
    _this.state = {};
    _this.getPaginationConfig = _this.getPaginationConfig.bind(_assertThisInitialized(_this));
    _this.getColumnsConfig = _this.getColumnsConfig.bind(_assertThisInitialized(_this));
    _this.getDataSource = _this.getDataSource.bind(_assertThisInitialized(_this));
    _this.columns = _this.getColumnsConfig();
    return _this;
  }

  _createClass(MamTable, [{
    key: "getPaginationConfig",
    value: function getPaginationConfig() {
      var pagination = this.props.pagination;

      if ((0, _isBoolean["default"])(pagination) && !pagination) {
        return pagination;
      }

      return Object.assign({
        showQuickJumper: true,
        showSizeChanger: true,
        pageSizeOptions: ['10', '20', '30', '40', '50']
      }, (0, _getComponentProps3["default"])(pagination));
    }
  }, {
    key: "getColumnsConfig",
    value: function getColumnsConfig() {
      var _this$props = this.props,
          serialNo = _this$props.serialNo,
          columns = _this$props.columns;
      var newColumns = columns.map(function (item) {
        var render = item.render;

        if (!('render' in item) && 'actions' in item && Array.isArray(item.actions)) {
          var actionsLen = item.actions.length;

          render = function render(text, record) {
            return item.actions.map(function (ac, i) {
              var name = ac.name,
                  onClick = ac.onClick,
                  popconfirm = ac.popconfirm;
              var btnNode = React.createElement("a", {
                key: name,
                onClick: onClick && !popconfirm ? function () {
                  return onClick(text, record);
                } : undefined
              }, name);

              if (popconfirm) {
                var confirm = popconfirm.confirm,
                    rest = __rest(popconfirm, ["confirm"]);

                btnNode = React.createElement(_popconfirm["default"], _extends({}, rest, {
                  onConfirm: confirm ? function () {
                    return confirm(record);
                  } : undefined,
                  key: name
                }), btnNode);
              }

              return [btnNode, i != actionsLen - 1 && React.createElement(_divider["default"], {
                key: "divider_".concat(i),
                type: "vertical"
              })];
            });
          };
        }

        return Object.assign(Object.assign({}, item), {
          render: render
        });
      });

      if (serialNo) {
        var _getComponentProps = (0, _getComponentProps3["default"])(serialNo),
            _getComponentProps$co = _getComponentProps.columnTitle,
            columnTitle = _getComponentProps$co === void 0 ? '序号' : _getComponentProps$co;

        var serialNoColumn = {
          title: columnTitle,
          dataIndex: 'serialNo',
          key: 'serialNo',
          width: 70
        };
        newColumns = [serialNoColumn].concat(_toConsumableArray(newColumns));
      }

      return newColumns;
    }
  }, {
    key: "getDataSource",
    value: function getDataSource() {
      var _this$props2 = this.props,
          serialNo = _this$props2.serialNo,
          _this$props2$dataSour = _this$props2.dataSource,
          dataSource = _this$props2$dataSour === void 0 ? [] : _this$props2$dataSour;
      var newData = dataSource;

      if (serialNo) {
        var _getComponentProps2 = (0, _getComponentProps3["default"])(serialNo),
            _getComponentProps2$s = _getComponentProps2.startIndex,
            startIndex = _getComponentProps2$s === void 0 ? 1 : _getComponentProps2$s;

        newData = dataSource.map(function (item, index) {
          return Object.assign(Object.assign({}, item), {
            serialNo: index + startIndex
          });
        });
      }

      return newData;
    }
  }, {
    key: "render",
    value: function render() {
      var rest = __rest(this.props, []);

      return React.createElement(_table["default"], _extends({}, rest, {
        columns: this.columns,
        dataSource: this.getDataSource(),
        pagination: this.getPaginationConfig()
      }));
    }
  }]);

  return MamTable;
}(React.Component);

exports["default"] = MamTable;
