// @uiw/react-native/lib/Tile/index.js
import React from 'react';
import { View, Text, StyleSheet, Dimensions, TouchableOpacity, } from 'react-native';
import Icon from '../Icon';
import TransitionImage from '../TransitionImage';
import { useTheme } from '@shopify/restyle';
const Tile = (props) => {
    const { imageSrc, icon, title, onPress, caption, activeOpacity, captionStyle, containerStyle, titleStyle, iconContainerStyle, contentContainerStyle, imageContainerStyle, titleNumberOfLines, imageProps = {}, width = Dimensions.get('window').width, height = width * 0.8, ImageComponent = TransitionImage, ...attributes } = props;
    const theme = useTheme();
    const styles = createStyles({
        titleColor: theme.colors.primary_text,
        captionColor: theme.colors.text,
    });
    return (<TouchableOpacity {...attributes} onPress={onPress} activeOpacity={activeOpacity} style={StyleSheet.flatten([
            {
                width: width,
                height: height,
            },
            containerStyle && containerStyle,
        ])}>
      <ImageComponent source={imageSrc} containerStyle={StyleSheet.flatten([styles.imageContainer, imageContainerStyle && imageContainerStyle])} style={{
            // ...StyleSheet.absoluteFillObject,
            alignItems: 'center',
            justifyContent: 'center',
            width,
            height: 200,
        }} {...imageProps}>
        <View style={StyleSheet.flatten([iconContainerStyle && iconContainerStyle])}>{icon && <Icon {...icon}/>}</View>
      </ImageComponent>

      <View style={[
            StyleSheet.flatten([styles.contentContainer, contentContainerStyle && contentContainerStyle]),
            { ...StyleSheet.absoluteFillObject },
        ]}>
        <Text testID="tileTitle" style={StyleSheet.flatten([styles.titleText, titleStyle && titleStyle])} numberOfLines={titleNumberOfLines}>
          {title}
        </Text>
        <Text testID="tileText" style={StyleSheet.flatten([styles.captionText, captionStyle && captionStyle])} numberOfLines={titleNumberOfLines}>
          {caption}
        </Text>
      </View>
    </TouchableOpacity>);
};
function createStyles({ titleColor, captionColor }) {
    return StyleSheet.create({
        imageContainer: {},
        titleText: {
            backgroundColor: 'rgba(0,0,0,0)',
            marginBottom: 5,
            color: titleColor,
        },
        captionText: {
            backgroundColor: 'rgba(0,0,0,0)',
            marginBottom: 5,
            color: captionColor,
        },
        contentContainer: {
            paddingTop: 15,
            paddingBottom: 5,
            paddingLeft: 15,
            paddingRight: 15,
        },
        iconContainer: {
            justifyContent: 'center',
            alignItems: 'center',
            alignSelf: 'center',
        },
    });
}
export default Tile;
