// @shopify/restyle/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.composeRestyleFunctions = exports.createRestyleComponent = exports.createRestyleFunction = exports.createTheme = exports.useResponsiveProp = exports.useRestyle = exports.useTheme = exports.ThemeContext = exports.ThemeProvider = exports.createText = exports.createBox = exports.createVariant = void 0;
__exportStar(require("./restyleFunctions"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./createVariant"), exports);
__exportStar(require("./createBox"), exports);
__exportStar(require("./createText"), exports);
var createVariant_1 = require("./createVariant");
Object.defineProperty(exports, "createVariant", { enumerable: true, get: function () { return __importDefault(createVariant_1).default; } });
var createBox_1 = require("./createBox");
Object.defineProperty(exports, "createBox", { enumerable: true, get: function () { return __importDefault(createBox_1).default; } });
var createText_1 = require("./createText");
Object.defineProperty(exports, "createText", { enumerable: true, get: function () { return __importDefault(createText_1).default; } });
var context_1 = require("./context");
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return context_1.ThemeProvider; } });
Object.defineProperty(exports, "ThemeContext", { enumerable: true, get: function () { return context_1.ThemeContext; } });
var useTheme_1 = require("./hooks/useTheme");
Object.defineProperty(exports, "useTheme", { enumerable: true, get: function () { return __importDefault(useTheme_1).default; } });
var useRestyle_1 = require("./hooks/useRestyle");
Object.defineProperty(exports, "useRestyle", { enumerable: true, get: function () { return __importDefault(useRestyle_1).default; } });
var useResponsiveProp_1 = require("./hooks/useResponsiveProp");
Object.defineProperty(exports, "useResponsiveProp", { enumerable: true, get: function () { return __importDefault(useResponsiveProp_1).default; } });
var createTheme_1 = require("./createTheme");
Object.defineProperty(exports, "createTheme", { enumerable: true, get: function () { return __importDefault(createTheme_1).default; } });
var createRestyleFunction_1 = require("./createRestyleFunction");
Object.defineProperty(exports, "createRestyleFunction", { enumerable: true, get: function () { return __importDefault(createRestyleFunction_1).default; } });
var createRestyleComponent_1 = require("./createRestyleComponent");
Object.defineProperty(exports, "createRestyleComponent", { enumerable: true, get: function () { return __importDefault(createRestyleComponent_1).default; } });
var composeRestyleFunctions_1 = require("./composeRestyleFunctions");
Object.defineProperty(exports, "composeRestyleFunctions", { enumerable: true, get: function () { return __importDefault(composeRestyleFunctions_1).default; } });
